rust/noosphere-cli/src/native/commands/auth.rs

```diff
@@ -120,62 +72,184 @@ IMPORTANT: You MUST sync to enable your gateway to recognize the authorization:
+            let ucan = Ucan::try_from(jwt.as_str())?;
+
+            // TODO: Maybe only consider Noosphere-related proofs here
+            // TODO: Maybe filter on proofs that specifically refer to the current sphere
+            let proofs = ucan
+                .proofs()
+                .clone()
+                .unwrap_or_default()
+                .into_iter()
+                .filter_map(|cid| Cid::try_from(cid.as_str()).ok().map(Link::from))
+                .collect::<Vec<Link<Jwt>>>();
+
+            if *ucan.issuer() == sphere_identity {
+                // TODO: Such an authorization ought not have any topical proofs,
+                // but perhaps we should verify that
+                authorization_roots.push(link.clone())
+            } else {
+                for proof in proofs {
+                    let items = match authorization_hierarchy.get_mut(&proof) {
+                        Some(items) => items,
+                        None => {
+                            authorization_hierarchy.insert(proof.clone(), Vec::new());
+                            authorization_hierarchy.get_mut(&proof).ok_or_else(|| {
+                                anyhow!(
+                                    "Could not access list of child authorizations for {}",
+                                    &proof
+                                )
+                            })?
+                        }
+                    };
+                    items.push(link.clone());
+                }
+            }
+        }
+
+        if as_json {
+            let mut hierarchy = BTreeMap::new();
+            for (proof, children) in authorization_hierarchy {
+                hierarchy.insert(
+                    proof.to_string(),
+                    children
+                        .iter()
+                        .map(|child| child.to_string())
+                        .collect::<Vec<String>>(),
+                );
+            }
+            let mut meta = BTreeMap::new();
+            for (link, (name, id, token)) in authorization_meta {
+                meta.insert(
+                    link.to_string(),
+                    json!({
+                        "name": name,
+                        "id": id,
+                        "token": token
+                    }),
+                );
+            }
+
+            let roots = authorization_roots
+                .into_iter()
+                .map(|root| root.to_string())
+                .collect::<Vec<String>>();
+
+            info!(
+                "{}",
+                serde_json::to_string_pretty(&json!({
+                    "sphere": sphere_identity,
+                    "roots": roots,
+                    "hierarchy": hierarchy,
+                    "meta": meta
+                }))?
+            );
+        } else {
+            info!(" ⌀ Sphere");
+            info!(" │ {sphere_identity}");
+            info!(" │");
+
+            fn draw_branch(
```

Comment:
Rust newbie question:  In other programming languages, i would have said the following. I'm not certain if it applies. Thoughts?


Generating a big method like this inside of another method seems odd. Should this live elsewhere?
Also, you have differing functionality using the same input (json vs ascii tree rendering). Is there a better programming construct to separate these, like an interface/trait?

Reply:
This is pretty fresh (written today), I wouldn't consider it a stylistic reference.

Similar response to "differing functionality" in this case.

As for factoring, there are two questions I would ask myself:

- Is there obvious headroom for generalization across branches?
- How frequently will I benefit from a generic interface?

Rust gives us traits, which are similar to (but not quite the same as) interfaces. Traits can be a powerful tool for defining generic behavior, but they come with non-zero cognitive and syntactical overhead in order to use them.

In this case, the two render modes are pretty different in nature (one uses `serde_json`, the other is all manual formatting). And, most branches are pretty small. Pulling out different branches into their own functions is a relatively easy, low-overhead way to clean it up, and you can even use the signature as an approximation for a generic interface.

If we reached a point where rendering was made to be pluggable, or at least had more than a small handful of modes, I might consider a more sophisticated / extensible approach using traits and generics.

After all, it's always nice to have the headroom to refactor into baroque abstractions down the road :sweat_smile:

Reply:
Another version of this might be: Is there benefit to making the Vec that contains these a Real Type™ and then putting those serialization options on it? My understanding is that rustaceans tend to prefer NOT to do this sort of thing.

Reply:
> Another version of this might be: Is there benefit to making the Vec that contains these a Real Type™ and then putting those serialization options on it? My understanding is that rustaceans tend to prefer NOT to do this sort of thing.

Definitely! That can be a nice way to encapsulate behavior. I would probably come down on: it's a stylistic choice in this case. But, preferences aside, I think it's a valid way to express things like this. If there is an obvious advantage to doing it that way compared to a few small functions, go for it!

To speculate, I imagine that one case where Rust programmers may avoid using a stateful struct is when they are dealing in a lot of shared data and want to keep overhead as low as possible (both resource- and cognition-wise). In this case, one likely needs to perform acrobatics with lifetimes; these can be harder to express when you have structs keeping references to shared data. The lifetimes usually can be expressed, but the effort to do it may serve as a friction. It is sometimes easier to express these cases as stateless functions.